lab_3/main.py

```diff
@@ -59,49 +95,97 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
     :param corpus: a tuple of sentences
     :return: a tuple of the encoded sentences
     """
-    pass
+    if not isinstance(storage, LetterStorage) or not isinstance(corpus, tuple):
+        return ()
+
+    def encode_word(word: str) -> Tuple[int, ...]:
+        return tuple(storage.get_id_by_letter(char) for char in word)
+
+    def encode_sentence(sent: tuple) -> Tuple[Tuple[int, ...], ...]:
+        return tuple(encode_word(word) for word in sent)
+
+    return tuple(encode_sentence(sentence) for sentence in corpus)
 
 
 # 6
 class NGramTrie:
 
     def __init__(self, n: int):
-        pass
+        self.size: int = n
+        self.n_grams: tuple = ()
+        self.n_gram_frequencies: Dict[Tuple[int, ...], int] = {}
+        self.n_gram_log_probabilities: Dict[Tuple[int, ...], float] = {}
 
     def fill_n_grams(self, encoded_text: tuple) -> int:
         """
         Extracts n-grams from the given sentence, fills the field n_grams
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not isinstance(encoded_text, tuple):
+            return 1
+
+        def word_n_grams(word: Tuple[int, ...]) -> Tuple[int, ...]:
+            for idx in range(len(word) - self.size + 1):
+                yield tuple(word[idx:idx + self.size])
+
+        def sent_n_grams(sent: tuple) -> Tuple[Tuple[int, ...]]:
+            for word in sent:
+                yield tuple(word_n_grams(word))
+
+        self.n_grams = tuple(tuple(sent_n_grams(sent)) for sent in encoded_text)
+
+        return 0
 
     def calculate_n_grams_frequencies(self) -> int:
         """
         Fills in the n-gram storage from a sentence, fills the field n_gram_frequencies
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not self.n_grams:
+            return 1
+
+        for sentence in self.n_grams:
+            for word in sentence:
+                for n_gram in word:
+                    self.n_gram_frequencies[n_gram]: int = self.n_gram_frequencies.get(n_gram, 0) + 1
+
+        return 0
 
     def calculate_log_probabilities(self) -> int:
         """
         Gets log-probabilities of n-grams, fills the field n_gram_log_probabilities
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not self.n_gram_frequencies:
+            return 1
+
+        for n_gram, freq in self.n_gram_frequencies.items():
+            n_gram_prob: int = sum(_freq for token, _freq in self.n_gram_frequencies.items() if n_gram[0] == token[0])
```

Comment:
why do we compare 0 indexes?

Reply:
Well.

1) Yeah, I know that there's no need to be obsessed with type annotations. The only reason of annotating everything was my boredom (and I planned to remove most of them later, so I did it now).

2) Okay, zero indices. Let's assume that we have the following bi-gram tuples:

`n_gram = n_gram_frequency.keys() = ( (0, 1), (0, 2), (0, 3), (1, 1), (1, 5), (1, 8), (2, 2), (2, 14) )`

Suppose that the first element of each tuple is `w0` and the second one is `w1`. We need to find out all `w1` that occur after `w0`. We iterate over all 2-gram tuples and take their first elements (cf. `n_gram[0]`), then we compare it with other tuples where the first element is also `w0`. 

Thus, *at each iteration* we have summed up frequency scores for *fixed* `w0` ∈ { `set(n_gram[0])` } and various `w1` ∈ { `set(n_gram[1])` | `n_gram[0] == w0` }.

I hope I answered your question.